plugin/storage/grpc/shared/grpc_client.go

```diff
@@ -0,0 +1,179 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shared
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/proto-gen/storage_v1"
+	"github.com/jaegertracing/jaeger/storage/dependencystore"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+// grpcClient implements shared.StoragePlugin and reads/writes spans and dependencies
+type grpcClient struct {
+	readerClient     storage_v1.SpanReaderPluginClient
+	writerClient     storage_v1.SpanWriterPluginClient
+	depsReaderClient storage_v1.DependenciesReaderPluginClient
+}
+
+// DependencyReader implements shared.StoragePlugin.
+func (c *grpcClient) DependencyReader() dependencystore.Reader {
+	return c
+}
+
+// SpanReader implements shared.StoragePlugin.
+func (c *grpcClient) SpanReader() spanstore.Reader {
+	return c
+}
+
+// SpanWriter implements shared.StoragePlugin.
+func (c *grpcClient) SpanWriter() spanstore.Writer {
+	return c
+}
+
+// GetTrace takes a traceID and returns a Trace associated with that traceID
+func (c *grpcClient) GetTrace(ctx context.Context, traceID model.TraceID) (*model.Trace, error) {
+	stream, err := c.readerClient.GetTrace(ctx, &storage_v1.GetTraceRequest{
+		TraceID: traceID,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "plugin error")
+	}
+
+	trace := model.Trace{}
+	for received, err := stream.Recv(); err != io.EOF; received, err = stream.Recv() {
+		if err != nil {
+			return nil, errors.Wrap(err, "grpc stream error")
+		}
+
+		for _, span := range received.Spans {
+			trace.Spans = append(trace.Spans, &span)
+		}
+	}
+
+	return &trace, nil
+}
+
+// GetServices returns a list of all known services
+func (c *grpcClient) GetServices(ctx context.Context) ([]string, error) {
+	resp, err := c.readerClient.GetServices(ctx, &storage_v1.GetServicesRequest{})
+	if err != nil {
+		return nil, errors.Wrap(err, "plugin error")
+	}
+
+	return resp.Services, nil
+}
+
+// GetOperations returns the operations of a given service
+func (c *grpcClient) GetOperations(ctx context.Context, service string) ([]string, error) {
+	resp, err := c.readerClient.GetOperations(ctx, &storage_v1.GetOperationsRequest{
+		Service: service,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "plugin error")
+	}
+
+	return resp.Operations, nil
+}
+
+// FindTraces retrieves traces that match the traceQuery
+func (c *grpcClient) FindTraces(ctx context.Context, query *spanstore.TraceQueryParameters) ([]*model.Trace, error) {
```

Comment:
Writing tests and I can see a test in the Cassandra test suit which passes nil as the query. Is that something that can actually happen? It'll cause this to panic below. Should we validate queries here rather than pushing that logic down to the grpc plugin?

Reply:
nil query doesn't make sense, imo. The query-service would never do that.